### PR TITLE
rest api: add simple query support

### LINKED ISSUE #91: Language facet behaviour (number of results)
Launch an empty query.
Chose a language in the language facet, for example "Russian" (91 results).
The language facet presents at this stage entries with a number in brackets supposed to correspond to the results both in Russian and in the other languages.
Chose the language "English" (3 results). We expect to obtain 3 results for which language is Russian and English.
The behaviour is different. We obtain all documents in Russian or in English. The number in brackets for English is no more 3 but 1269.

### LINKED ISSUE #755: search: problem with brackets [ ] in the query
**Describe the issue**
search: there is problem when the query contains brackets [ ].
Brackets seems to be used by Elasticsearch (?) and there is an error in the backend. 

**To Reproduce**
1. Let's say we look for the title "100 [cent] énigmes pour réussir en maths : 10-11 ans"   (document ID 1189 / ISBN 9782725618296) 
1. Go to ilsdev
1. In the public interface, do a search on "100 [cent] énigmes pour réussir en maths : 10-11 ans"
1. You'll get a message "Something bad happened; please try again later. "
1. do another search on "100 cent énigmes pour réussir en maths : 10-11 ans"
1. You'll see a result.

**Expected behavior**
The patron should get a result even he uses backets in the query.

**Software version**
  * Tested on ilsdev
  * rero-ils https://github.com/rero/rero-ils/commit/5bc39afd322577d05e0793b7e6ee1b0717a8b065  & rero-ils-ui https://github.com/rero/rero-ils-ui/commit/78f78f58311f97abd731b363016b19ab8ca9c791

### PR DESCRIPTION
**Pull request by jma**
The current elasticsearch query string used by the REST API is powerful
but should not be used for search boxes. It raises exceptions when
the query syntax contains errors. As described in the elaticsearch
documentation, a simple query string should be used. A new http query
optional parameters has been added and can be specified as follows:
`&simple=1`.
For example: `https://ils.rero.ch/global/search/?q=potter&simple=1`.

When the simple query syntax is chosen, the default boolean operator is
`AND`. Except this parameter, nothing has been modified.

* Adds new type of aggregation filter to perform a AND boolean operator
between the terms in the same aggregation. All aggregations filters
use now AND boolean operator.
* Updates elasticsearch mappings to enhance the search engine quality.
* Adds a new REST API list records operator to use a `simple_query_string`
instead of a `query_string`. The simple query string preform an AND
boolean operator by default.
* Adds search query tests.
* Adds missing utils test.
* Renames `global_lowercase_asciifolding` elasticsearch analyzer to
`default`. This makes the rero-ils custom analyzer to be the default for
all elasticsearch `text` fields. All elasticsearch mappings has been
simplified.
* Creates a new `custom_keyword` analyzer.
* Creates a custom elasticsearch image with the icu analysis plugin
(https://www.elastic.co/guide/en/elasticsearch/plugins/current/analysis-icu.html).
* Closes #755.
* Closes #91.

## Required by

- rero-ils-ui: https://github.com/rero/rero-ils-ui/pull/265

## Why are you opening this PR?

- Implements: https://tree.taiga.io/project/rero21-reroils/us/1440
- Fixes: TODO: list of issues

## How to test?

- Execute a REST API search query with multiple words by adding a `&simple=1` option parameter. The number of results should be less than without the param. Try also query that causes syntax errors.

__Note__: see (https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html#simple-query-string-syntax) for the supported syntax documentation.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

**Comment by benerken** [comment]
Tests of simple search on ilsdev:

1.  Search on a series (ex: Que sais-je ?) 

Que sais-je  => returns 174 records
but
Que sais-je ? => returns 0

There is an encoding of the question mark (?) to %3F. You can see it 
in the URL after the search. See below:

image

More later....

**Comment by jma** [comment]
Other problems to solve or tests to write:

- maison blanche == Maison Blanche == Maison + Blanche == MAISON BLANCHE
- éléphant == elephant	21 == éléphant_ (with space)	 == elephant_ (with space) == ELEPHANT
- ABBE == abbe == abbé	17 == ABBé == abbé_
- müller == muller == mueller (if possible)
- saute == sautent	== saute* == sautez
- saute* -(sautent | sautez)
- pelage + (blanc | tigre)
- "écrivain espagnol"
- auteur enfant == auteur enfants
- Notes sur l'affaire Dominici / Essai sur le caractère des personnages
- Theorien der Sozialpsychologie. Bd. 3, Motivations- und Informationsverarbeitungstheorien / mit Beitr. von Mark W. Baldwin ... [et al.]
- Que sais-je ?
- 500 [cinq cents] modèles au point de croix
- économie == economie

**Comment by jma** [comment]
All problems has been fixed and tests has been written. Some of them cannot be addressed such as économie == economie due to the multi lingual support.

File: rero_ils/query.py [hunk 1]
```diff
@@ -135,14 +153,25 @@ def search_factory(self, search, query_parser=None):
     """
     def _default_parser(qstr=None, query_boosting=[]):
         """Default parser that uses the Q() from elasticsearch_dsl."""
+        query_type = 'query_string'
+        default_operator = 'OR'
+        if request.args.get('simple'):
```
Comment (sebastiendeleze) [inline rero_ils/query.py:158]: Why operator is set from *simple* argument. I think it's not fully related and should be passed as another parameter.

Comment (jma) [reply rero_ils/query.py:158]: I do not want to have two parameters even if you are again right, it is cleaner.

**Comment by jma** [comment]
> * Autosuggest display is not optimal when using boolean operators:
>   image
> * Œ,œ
>   
>   * [Œil or œil](https://ilsdev.test.rero.ch/global/search/documents?q=%C5%92il&page=1&size=10), 114 documents
>   * [oeil](https://ilsdev.test.rero.ch/global/search/documents?q=oeil&page=1&size=10), 71 documents
>   * same for ae
>   * ss, ü, ö are working fine
> * The minus to negate a token from the search does not work: [example](https://ilsdev.test.rero.ch/global/search/documents?q=christian%20philipp%20-autofocus&page=1&size=10)
> * Detail: it could be convenient to be able to search by patron birth year (tokenizer for birth date). Currently, full input ["1961-11-24"](https://ilsdev.test.rero.ch/professional/records/patrons?q=1961-11-24&page=1&size=10) is necessary. Ideally, input ["1961"](https://ilsdev.test.rero.ch/professional/records/patrons?q=1961&page=1&size=10) is working as well

All has been addressed.

**Pull request merged**